### ISSUE TITLE
Make Gitlab as an option instead of Github

### ISSUE THREAD

**Issue by Repooc** [p3lim/addon-packager-proxy#12]
Like to have Gitlab as an option as we use that for everything including ticket system and prefer not to move back to github

**Comment by p3lim** [comment]
If GitLab has webhooks support, give me a typical dump of a payload and I'll take a look.
If not then you're out of luck.

**Comment by Repooc** [comment]
```

HEADERS

Total-Route-Time: 0
Cf-Ipcountry: CA
Content-Length: 2853
X-Gitlab-Event: Push Hook
Cf-Ray: 308b698be3c0552e-ORD
Connect-Time: 0
Content-Type: application/json
Cf-Connecting-Ip: 2607:5300:0060:14a7:0000:0000:0000:0001
Connection: close
Accept-Encoding: gzip
X-Request-Id: dc1107c1-8ed8-462a-a3ad-05db50b8927c
Via: 1.1 vegur
Cf-Visitor: {"scheme":"http"}
Host: requestb.in
```
and the data here
```
{"object_kind":"push","event_name":"push","before":"f1b930147ba20a9942f2e9080e6788fba5dbaade","after":"1a6b44c2113d3cbac97ac85d7f5f5bcb9039cf29","ref":"refs/heads/master","checkout_sha":"1a6b44c2113d3cbac97ac85d7f5f5bcb9039cf29","message":null,"user_id":14,"user_name":"Repooc","user_email":"tjcooper1982@gmail.com","user_avatar":"http://www.gravatar.com/avatar/b4b2d920375fb1d1d6df5dce18c5fb4a?s=80\u0026d=identicon","project_id":23,"project":{"name":"Shadow and Light","description":"ElvUI Plugin with a lot of additional features","web_url":"http://git.tukui.org/repooc/elvui-shadowandlight","avatar_url":"http://git.tukui.org/uploads/project/avatar/23/SLE_Banner.png","git_ssh_url":"git@git.tukui.org:repooc/elvui-shadowandlight.git","git_http_url":"http://git.tukui.org/repooc/elvui-shadowandlight.git","namespace":"repooc","visibility_level":20,"path_with_namespace":"repooc/elvui-shadowandlight","default_branch":"master","homepage":"http://git.tukui.org/repooc/elvui-shadowandlight","url":"git@git.tukui.org:repooc/elvui-shadowandlight.git","ssh_url":"git@git.tukui.org:repooc/elvui-shadowandlight.git","http_url":"http://git.tukui.org/repooc/elvui-shadowandlight.git"},"commits":[{"id":"1a6b44c2113d3cbac97ac85d7f5f5bcb9039cf29","message":"Altered item populate function for loc bar dropdown to work like toolbars.lua should now more accurately detect items\n","timestamp":"2016-11-27T20:56:52+03:00","url":"http://git.tukui.org/repooc/elvui-shadowandlight/commit/1a6b44c2113d3cbac97ac85d7f5f5bcb9039cf29","author":{"name":"Darth Predator","email":"darthpred@gmail.com"},"added":[],"modified":["ElvUI_SLE/modules/minimap/locationbar.lua"],"removed":[]},{"id":"f730bda763e172dee5e7b16e10b59677da78a0d1","message":"Maybe looks better???\n","timestamp":"2016-11-26T20:34:27-05:00","url":"http://git.tukui.org/repooc/elvui-shadowandlight/commit/f730bda763e172dee5e7b16e10b59677da78a0d1","author":{"name":"Repooc","email":"tjcooper1982@gmail.com"},"added":[],"modified":["ElvUI_SLE/modules/Armory/InspectArmory/InspectArmory.lua"],"removed":[]},{"id":"f1b930147ba20a9942f2e9080e6788fba5dbaade","message":"Move inspect buttons down a bit\n","timestamp":"2016-11-26T19:12:13-05:00","url":"http://git.tukui.org/repooc/elvui-shadowandlight/commit/f1b930147ba20a9942f2e9080e6788fba5dbaade","author":{"name":"Repooc","email":"tjcooper1982@gmail.com"},"added":[],"modified":["ElvUI_SLE/modules/Armory/InspectArmory/InspectArmory.lua"],"removed":[]}],"total_commits_count":3,"repository":{"name":"Shadow and Light","url":"git@git.tukui.org:repooc/elvui-shadowandlight.git","description":"ElvUI Plugin with a lot of additional features","homepage":"http://git.tukui.org/repooc/elvui-shadowandlight","git_http_url":"http://git.tukui.org/repooc/elvui-shadowandlight.git","git_ssh_url":"git@git.tukui.org:repooc/elvui-shadowandlight.git","visibility_level":20}}
```

**Comment by Repooc** [comment]
Is that what you are looking for?
Only used pushed settings but here is the list of settings available for the webhooks.
![](https://i.imgur.com/l42wP16.png)

**Comment by p3lim** [comment]
Thanks, I'll take a look at it when I get the chance.

**Comment by p3lim** [comment]
Alright, haven't tested it yet but I drycoded support for GitLab webhooks.
I added Bitbucket as well, but I'm not going to bother with Gogs.io unless someone asks for it.

Some things to note though, GitLab doesn't encrypt their secret, it's sent in plain text as a header, and Bitbucket doesn't even have the option for a secret.
This is very bad security practice, as the webhooks from GitLab is subject to MITM attacks, and with Bitbucket it doesn't matter as the packager would just accept any requests.
